src/lib/Refiner.ml

```diff
@@ -605,12 +605,15 @@ struct
     index ix
 
   let let_ tac_def (nm_x, (tac_bdy : T.var -> T.bchk_tac)) : T.bchk_tac =
-    fun tp ->
+    fun goal ->
     let* tdef, tp_def = tac_def in
+    EM.with_pp @@ fun ppenv ->
     let* vdef = EM.lift_ev @@ Nbe.eval tdef in
-    T.let_ tp_def vdef nm_x @@ fun x ->
-    let+ tbdy = tac_bdy x tp in
-    S.Let (tdef, tbdy)
+    let* tbdy =
+      T.abstract (D.Sub (tp_def, Cof_tl.top, D.const_tm_clo vdef)) nm_x @@ fun var ->
+      tac_bdy var goal
```

Comment:
In fact, I think that this is not quite right. (It works but it isn't yet right).

Reply:
The problem is that it seems like we can't actually push in the boundary condition for this style of rule. I need to look at it on paper before I resume. But I'm way too busy today.

Reply:
Oh wait, I think it's ok because the binding has the singleton type!